README.rst
droprowsbyposition
------------------

Workbench module that drops specified rows.

Developing
----------

First, get up and running:

#. ``python3 ./setup.py test`` # to test Python
#. ``npm install && npm test`` # to test JavaScript

To add a feature on the Python side:

#. Write a test in ``test_droprowsbyposition.py``
#. Run ``python3 ./setup.py test`` to prove it breaks
#. Edit ``droprowsbyposition.py`` to make the test pass
#. Run ``python3 ./setup.py test`` to prove it works
#. Commit and submit a pull request

To add a feature on the JavaScript side:

#. Write a test in ``support.test.js``
#. Run ``npm test`` to prove it breaks
#. Edit ``support.mjs`` to make the test pass
#. Run ``npm test`` to prove it works
#. Commit and submit a pull request

To develop continuously on Workbench:

#. Check out the columnchart repository in a sibling directory to your checked-out Workbench code.
#. Start Workbench with ``bin/dev start`` 
#. In a separate tab in the Workbench directory, run ``pipenv run ./manage.py develop-module droprowsbyposition``
#. Edit this code; the module will be reloaded in Workbench immediately. In the Workbench website, modify parameters to execute the reloaded code.

For full instructions, see [developing your own modules](https://github.com/CJWorkbench/cjworkbench/wiki/Creating-A-Module)


CHANGELOG.md
2021-03-02.01
~~~~~~~~~~~~~

* Warn when user enters a "backwards" range, like "3-1".


test/support.test.js
const expect = require('chai').expect
const { stringToMask, maskToString, orMasks, composeMasks, addSelectedRows } = require('../support')

function MASK(...args) {
  return new Uint8Array(args)
}

describe('support', function () {
  describe('stringToMask', function () {
    it('should handle single numbers', function () {
      expect(stringToMask('1, 3, 4'))
        .to.deep.equal(MASK(1, 0, 1, 1))
    })

    it('should handle empty', function () {
      expect(stringToMask(''))
        .to.deep.equal(MASK())
    })

    it('should handle ranges', function () {
      expect(stringToMask('1-2, 4-5'))
        .to.deep.equal(MASK(1, 1, 0, 1, 1))
    })

    it('should handle length-1 ranges', function () {
      expect(stringToMask('1-1, 3-3'))
        .to.deep.equal(MASK(1, 0, 1))
    })

    it('should do nothing for zero- or negative-length ranges', function () {
      expect(stringToMask('2-1, 5-3'))
        .to.deep.equal(MASK())
    })

    it('should allow overlapping ranges', function () {
      expect(stringToMask('1-4, 2-5'))
        .to.deep.equal(MASK(1, 1, 1, 1, 1))
    })

    it('should ignore single-value zero', function () {
      expect(stringToMask('0'))
        .to.deep.equal(MASK())
    })

    it('should ignore zero in a range', function () {
      expect(stringToMask('0-2'))
        .to.deep.equal(MASK(1, 1))
    })

    it('should ignore too-many-values range', function () {
      expect(stringToMask('1-3-4'))
        .to.deep.equal(MASK())
    })

    it('should round down when given fractions', function () {
      expect(stringToMask('1.2, 3.1-4.8'))
        .to.deep.equal(MASK(1, 0, 1, 1))
    })

    it('should ignore non-numbers', function () {
      // Changing the retval here wouldn't be evil -- it's basically undefined
      // behavior. Mainly, we're testing that the output is reasonable and the
      // function doesn't crash.
      expect(stringToMask('1, foo-2, 3ef'))
        .to.deep.equal(MASK(1, 0, 1))
    })

    it('should limit range to a maximum', function () {
      expect(stringToMask('2-99', { maxMaskLength: 5 }))
        .to.deep.equal(MASK(0, 1, 1, 1, 1))
    })

    it('should limit single number to a maximum', function () {
      expect(stringToMask('2, 99', { maxMaskLength: 5 }))
        .to.deep.equal(MASK(0, 1))
    })

    it('should be speedy with big numbers', function () {
      expect(stringToMask('1-999999999999')).to.have.length(2000000)
    })
  })

  describe('maskToString', function () {
    it('should output single values', function () {
      expect(maskToString(MASK(1, 0, 1))).to.equal('1, 3')
    })

    it('should output ranges', function () {
      expect(maskToString(MASK(0, 1, 1, 0, 1, 1))).to.equal('2-3, 5-6')
    })
  })

  describe('orMasks', function () {
    it('should OR two masks', function () {
      expect(orMasks(MASK(1, 0, 0, 1), MASK(1, 1, 0, 0)))
        .to.deep.equal(MASK(1, 1, 0, 1))
    })

    it('should OR two masks when mask1 is longer', function () {
      expect(orMasks(MASK(1, 0, 0, 1), MASK(1, 1)))
        .to.deep.equal(MASK(1, 1, 0, 1))
    })

    it('should OR two masks when mask2 is longer', function () {
      expect(orMasks(MASK(0, 1), MASK(1, 0, 0, 1)))
        .to.deep.equal(MASK(1, 1, 0, 1))
    })
  })

  describe('composeMasks', function () {
    it('should compose masks', function () {
      // 1. Delete rows 1 and 4. Remaining rows: 2, 3, 5, 6, 7, ...
      // 2. Delete rows 2 and 3. Remaining rows: 2, 6, 7, ... (1, 3-5 deleted)
      expect(composeMasks(MASK(1, 0, 0, 1), MASK(0, 1, 1)))
        .to.deep.equal(MASK(1, 0, 1, 1, 1))
    })

    it('should add second-mask deletes after first mask ends', function () {
      // 1. Delete row 1. Remaining rows: 2, 3, 4, 5, ...
      // 2. Delete row 2. Remaining rows: 2, 4, 5, ... (1, 3 deleted)
      expect(composeMasks(MASK(1), MASK(0, 1)))
        .to.deep.equal(MASK(1, 0, 1))
    })

    it('should add first-mask deletes after second mask ends', function () {
      // 1. Delete row 3. Remaining rows: 1, 2, 4, 5, ...
      // 2. Delete row 1. Remaining rows: 2, 4, 5, ... (1, 3 deleted)
      expect(composeMasks(MASK(0, 0, 1), MASK(1)))
        .to.deep.equal(MASK(1, 0, 1))
    })

    it('should work when mask1[0] and mask[2] are false', function () {
      // 1. Delete rows 2, 4. Remaining rows: 1, 3, 5, 6, 7, 8, 9, ...
      // 2. Delete rows 2, 4. Remaining rows: 1, 5, 7, 8, 9, ... (2-4, 6 deleted)
      expect(composeMasks(MASK(0, 1, 0, 1), MASK(0, 1, 0, 1)))
        .to.deep.equal(MASK(0, 1, 1, 1, 0, 1))
    })

    it('should work when mask1 is all at beginning', function () {
      // 1. Delete rows 1-3. Remaining rows: 4, 5, 6, 7, 8, 9, ...
      // 2. Delete rows 2-3. Remaining rows: 4, 7, 8, 9, ... (1-3, 5-6 deleted)
      expect(composeMasks(MASK(1, 1, 1), MASK(0, 1, 1)))
        .to.deep.equal(MASK(1, 1, 1, 0, 1, 1))
    })

    it('should work when mask2 is all at beginning', function () {
      // 1. Delete rows 2-3. Remaining rows: 1, 4, 5, 6, 7, 8, 9, ...
      // 2. Delete rows 1-3. Remaining rows: 6, 7, 8, 9, ... (1-5 deleted)
      expect(composeMasks(MASK(0, 1, 1), MASK(1, 1, 1)))
        .to.deep.equal(MASK(1, 1, 1, 1, 1))
    })

    it('should work when mask2 is all before mask1', function () {
      // 1. Delete rows 4-5. Remaining rows: 1, 2, 3, 6, 7, 8, ...
      // 2. Delete rows 1-2. Remaining rows: 3, 6, 7, 8, ... (1-2, 4-5 deleted)
      expect(composeMasks(MASK(0, 0, 0, 1, 1), MASK(1, 1)))
        .to.deep.equal(MASK(1, 1, 0, 1, 1))
    })
  })

  describe('addSelectedRows', function () {
    it('should OR when adding deletions to input', function () {
      expect(addSelectedRows({ rows: '8-9' }, '2-3', true))
        .to.deep.equal({ rows: '2-3, 8-9' })
    })

    it('should COMPOSE when adding deletions to output', function () {
      expect(addSelectedRows({ rows: '1-3' }, '2-3', false))
        .to.deep.equal({ rows: '1-3, 5-6' })
    })

    it('should convert from schema-v0 to schema-v1 before compose', function () {
      expect(addSelectedRows({ rows: '', first_row: 1, last_row: 3 }, '2-3', false))
        .to.deep.equal({ first_row: 0, last_row: 0, rows: '1-3, 5-6' })
    })

    it('should be speedy with big numbers', function () {
      expect(addSelectedRows({ rows: '999999' }, '234012', false))
        .to.deep.equal({ rows: '234012, 999999' })
    })

    it('should nix excessively-large numbers', function () {
      expect(addSelectedRows({ rows: '999999999' }, '234012', false))
        .to.deep.equal({ rows: '234012' })
    })
  })
})


test_droprowsbyposition.py
import unittest
import pandas as pd
from pandas.testing import assert_frame_equal
from droprowsbyposition import migrate_params, render
from cjwmodule.testing.i18n import i18n_message


class TestMigrateParams(unittest.TestCase):
    def test_v0_with_first_row_and_last_row(self):
        # droprowsbyposition got some behavior prior to our migrate_params()
        # idea. So it actually has two logic paths....
        self.assertEqual(
            migrate_params(
                {
                    "rows": "",
                    "first_row": 2,
                    "last_row": 54,
                }
            ),
            {
                "rows": "2-54",
            },
        )

    def test_v0_with_rows(self):
        self.assertEqual(
            migrate_params(
                {
                    "rows": "1-23",
                    # first_row and last_row are 'backup' values, only for when 'rows'
                    # is not set. So they'll be dropped.
                    "first_row": 2,
                    "last_row": 54,
                }
            ),
            {
                "rows": "1-23",
            },
        )

    def test_v1(self):
        self.assertEqual(
            migrate_params(
                {
                    "rows": "1-23",
                }
            ),
            {
                "rows": "1-23",
            },
        )


class TestDropRowsByPosition(unittest.TestCase):
    def setUp(self):
        # Test data includes some partially and completely empty rows because
        # this tends to freak out Pandas
        self.table = pd.DataFrame(
            {
                "A": [1, 2, 3, 4],
                "B": [2, 3, None, 5],
                "C": [None, None, None, None],
            }
        )

    def test_default_params_NOP(self):
        out = render(self.table, {"rows": ""})
        assert_frame_equal(out, self.table)

    def test_empty_input_NOP(self):
        # https://www.pivotaltracker.com/n/projects/2132449/stories/161945860
        out = render(pd.DataFrame(), {"rows": ""})
        self.assertTrue(out.empty)

    def test_drop_one_row(self):
        out = render(self.table, {"rows": "2"})
        assert_frame_equal(
            out,
            pd.DataFrame(
                {
                    "A": [1, 3, 4],
                    "B": [2, None, 5],
                    "C": [None, None, None],
                }
            ),
        )

    def test_drop_first_rows(self):
        out = render(self.table, {"rows": "1-2"})
        assert_frame_equal(
            out,
            pd.DataFrame(
                {
                    "A": [3, 4],
                    "B": [None, 5],
                    "C": [None, None],
                }
            ),
        )

    def test_drop_last_rows(self):
        out = render(self.table, {"rows": "3-4"})
        assert_frame_equal(
            out,
            pd.DataFrame(
                {
                    "A": [1, 2],
                    "B": [2.0, 3.0],
                    "C": [None, None],
                }
            ),
        )

    def test_drop_multiple_interval_ranges(self):
        out = render(self.table, {"rows": "1, 3-4"})
        assert_frame_equal(
            out,
            pd.DataFrame(
                {
                    "A": [2],
                    "B": [3.0],
                    "C": [None],
                }
            ),
        )

    def test_drop_overlapping_interval_ranges(self):
        result = render(pd.DataFrame({"A": [1, 2, 3, 4, 5]}), {"rows": "1-2,2"})
        assert_frame_equal(result, pd.DataFrame({"A": [3, 4, 5]}))

    def test_drop_everything(self):
        out = render(self.table, {"rows": "1-4"})
        expected = pd.DataFrame({"A": [1], "B": [1.0], "C": [None]})[0:0]
        # Let's assert the same thing, three different ways...:
        self.assertEqual(len(out), 0)
        self.assertTrue(out.empty)
        assert_frame_equal(out, expected)

    def test_zero_gives_error(self):
        out = render(self.table, {"rows": "0-1"})
        self.assertEqual(
            out, i18n_message("badParam.rows.invalidRange", {"value": "0-1"})
        )

    def test_backwards_range_gives_error(self):
        out = render(self.table, {"rows": "3-1"})
        self.assertEqual(
            out, i18n_message("badParam.rows.backwardsRange", {"value": "3-1"})
        )

    def test_drop_after_end(self):
        out = render(self.table, {"rows": "5-8"})
        assert_frame_equal(
            out,
            pd.DataFrame(
                {
                    "A": [1, 2, 3, 4],
                    "B": [2, 3, None, 5],
                    "C": [None, None, None, None],
                }
            ),
        )

    def test_remove_unused_categories(self):
        result = render(
            pd.DataFrame({"A": ["a", "b", "c", "d"]}, dtype="category"), {"rows": "1-2"}
        )
        assert_frame_equal(result, pd.DataFrame({"A": ["c", "d"]}, dtype="category"))

    def test_remove_unused_categories_for_empty_row(self):
        result = render(
            pd.DataFrame({"A": ["a", "b", "c", "d"]}, dtype="category"), {"rows": "1-4"}
        )
        assert_frame_equal(result, pd.DataFrame({"A": []}, dtype="category"))


support.js
const DefaultOptions = {
  maxMaskLength: 2000000 // bound memory/CPU usage
}

// Internal logic is done using Uint8Array. It's almost as easy to use as an
// Array of booleans, and it's much faster/smaller. One can imagine smaller or
// faster data structures, but they'd be more complex.
//
// And yes, we need the space. With 1M entries, an Array of boolean costs
// 27MB more than a Uint8Array.
//
// $ node
// > a = []; for (let i = 0; i < 1000000; i++) { a[i] = (i % 2 === 1) }
// true
// > process.memoryUsage().
// { rss: 60563456,
//   heapTotal: 38899712,
//   heapUsed: 34884296,
//   external: 8613 }
// > process.exit()
// $ node
// > a = new Uint8Array(1000000); for (let i = 0; i < 1000000; i++) { a[i] = (i % 2 === 1) }
// true
// > process.memoryUsage()
// { rss: 31313920,
//   heapTotal: 7684096,
//   heapUsed: 5291464,
//   external: 1008688 }

function stringToMask (str, options) {
  if (!options) options = {}
  options = { ...DefaultOptions, ...options }
  const { maxMaskLength } = options

  const spans = str.split(/,/g) // split by comma

    // Parse "A-B" into [A, B] numeric ranges, zero-indexed.
    .map(s => s.split('-')
      .map(s2 => parseInt(s2.trim(), 10) - 1)
    )
    .filter(s => s.length === 1 || s.length === 2) // ignore wrong-length spans
    .map(s => s.length === 1 ? [ s, s ] : s)       // turn `A` into `[A, A]`
    .filter(([a, b]) => !isNaN(a) && !isNaN(b))    // nix parse errors
    .map(([a, b]) => [ Math.max(0, a), Math.min(maxMaskLength - 1, b) ]) // bound
    .filter(([a, b]) => a <= b)                    // nix 0- or negative-length spans

  // Output length is whatever the maximum number in the values/ranges.
  const len = spans.reduce((agg, s) => Math.max(agg, s[1] + 1), 0)

  const mask = new Uint8Array(len)

  for (const [ begin, end ] of spans) {
    for (let i = begin; i <= end; i++) {
      mask[i] = 1
    }
  }

  return mask
}

/**
 * OR two masks of booleans.
 *
 * Why: if we already deleted for 2 and we're asked to delete row 2 again, then
 * OR is the operation we need: its output will be, 'Delete row 2'.
 *
 * @param mask1 One mask
 * @param mask2 Another mask
 */
function orMasks (mask1, mask2) {
  const shortMask = mask1.length < mask2.length ? mask1 : mask2
  const longMask = shortMask === mask1 ? mask2 : mask1
  const shortLen = shortMask.length

  const mask = new Uint8Array(longMask)
  for (let i = 0; i < shortLen; i++) {
    mask[i] = mask[i] || shortMask[i]
  }

  return mask
}

/**
 * Apply mask2 on top of the output of mask1.
 *
 * Why: if we deleted row 2 and are asked to delete rows 1-3 of the output of
 * that first delete operation, then COMPOSE is the operation we need: its
 * output will be, 'Delete rows 1-4'.
 *
 * Order of arguments matters:
 *
 *     compose(2, 1-3) => 1-4
 *     compose(1-3, 2) => 1-3, 5
 *
 * @param mask1 The original mask
 * @param mask2 A mask to apply to the output from applying mask1
 */
function composeMasks (mask1, mask2) {
  const len1 = mask1.length
  const len2 = mask2.length

  // Output length is hard to guess. Let's just double the RAM and truncate.
  const mask = new Uint8Array(len1 + len2)
  // Walk through the elements of `mask` (with pointer j). Delete according to
  // mask1 and mask2, which we iterate over using i1 and i2.
  let j = 0
  for (let i1 = 0, i2 = 0; i1 < mask1.length || i2 < mask2.length; j++) {
    if (i1 < len1 && mask1[i1]) {
      // Element is deleted, according to mask1. Mark it as such before even
      // checking against mask1.
      mask[j] = true
      i1++
    } else {
      // Apply mask2 and move forward
      mask[j] = i2 < len2 && mask2[i2]
      i2++
      i1++
    }
  }

  // Truncate so the last value is `true`
  return new Uint8Array(mask.slice(0, j))
}

function maskToString (mask) {
  const len = mask.length
  const spans = []

  let spanStart = null
  for (let i = 0; i < len; i++) {
    const bit = mask[i]
    if (bit && spanStart === null) {
      spanStart = i
    } else if (!bit && spanStart !== null) {
      spans.push([ spanStart, i - 1 ])
      spanStart = null
    }
  }
  if (spanStart !== null) {
    spans.push([ spanStart, len - 1 ])
  }

  return spans
    .map(span => span[0] === span[1] ? String(span[0] + 1) : `${span[0] + 1}-${span[1] + 1}`)
    .join(', ')
}

/**
 * Build changed params to incorporate the new selectedRows.
 *
 * @param oldParams Old Object params.
 * @param addRows String of new row numbers, e.g. '1-5, 8'
 * @param fromInput Boolean: true means the user selected addRows on the input
 *                  table; true means the user selected addRows on the output
 *                  table.
 */
function addSelectedRows (oldParams, addRows, fromInput) {
  const ret = {}
  let oldRows

  if (oldParams.first_row && oldParams.last_row && !oldParams.rows) {
    ret.first_row = 0
    ret.last_row = 0
    oldRows = [ oldParams.first_row, oldParams.last_row ].join('-')
  } else {
    oldRows = oldParams.rows || ''
  }

  const oldMask = stringToMask(oldRows)
  const addMask = stringToMask(addRows)
  let newMask
  if (fromInput) {
    newMask = orMasks(oldMask, addMask)
  } else {
    newMask = composeMasks(oldMask, addMask)
  }

  ret.rows = maskToString(newMask)

  return ret
}

module.exports = {
  stringToMask,
  orMasks,
  composeMasks,
  maskToString,
  addSelectedRows
}


droprowsbyposition.py
import re
from typing import Any, Dict, Tuple
import numpy as np
import pandas as pd
from cjwmodule import i18n

commas = re.compile(r"\s*,\s*")
numbers = re.compile(r"(?P<first>[1-9]\d*)(?:-(?P<last>[1-9]\d*))?")


class RangeFormatError(ValueError):
    def __init__(self, value):
        self.value = value

    @property
    def i18n_message(self) -> i18n.I18nMessage:
        return i18n.trans(
            "badParam.rows.invalidRange",
            'Rows must look like "1-2", "5" or "1-2, 5"; got "{value}"',
            {"value": self.value},
        )


class BackwardsRangeError(ValueError):
    def __init__(self, value):
        self.value = value

    @property
    def i18n_message(self) -> i18n.I18nMessage:
        return i18n.trans(
            "badParam.rows.backwardsRange",
            'Row ranges must increase like "1-3", not "3-1"; got "{value}"',
            {"value": self.value},
        )


def parse_interval(s: str) -> Tuple[int, int]:
    """
    Parse a string 'interval' into a tuple

    >>> parse_interval('1')
    (0, 1)
    >>> parse_interval('1-3')
    (0, 2)
    >>> parse_interval('5')
    (4, 4)
    >>> parse_interval('hi')
    Traceback (most recent call last):
        ...
    RangeFormatError: Rows must look like "1-2", "5" or "1-2, 5"; got "hi"
    >>> parse_interval('12-10')
    Traceback (most recent call last):
        ...
    BackwardsRangeError: Row ranges must increase like "1-3", not "3-1"; got "12-10"
    """
    match = numbers.fullmatch(s)
    if not match:
        raise RangeFormatError(s)

    first = int(match.group("first"))
    last = int(match.group("last") or first)
    if first > last:
        raise BackwardsRangeError(s)
    return (first - 1, last)


def parse_mask(rows: str, n_rows: int) -> np.array:
    """Parse 'rows', or raise RangeFormatError or BackwardsRangeError"""
    mask = np.zeros(n_rows, np.bool)
    for s in commas.split(rows.strip()):
        if s:
            begin, end = parse_interval(s)  # raise RangeFormatError/BackwardsRangeError
            mask[begin:end] = True
    return mask


def render(table, params):
    try:
        mask = parse_mask(params["rows"], len(table))
    except (RangeFormatError, BackwardsRangeError) as err:
        return err.i18n_message

    if table.empty:
        # https://www.pivotaltracker.com/n/projects/2132449/stories/161945860
        return table

    table = table[~mask]
    table.reset_index(drop=True, inplace=True)
    for column in table.columns:
        series = table[column]
        if hasattr(series, "cat"):
            series.cat.remove_unused_categories(inplace=True)
    return table


def _migrate_params_v0_to_v1(params):
    """
    v0: optional "first_row" and "last_row" were fallback if "rows" empty.

    v1: always use "rows".

    (droprowsbyposition had some special logic and special visible_if rules
    because we changed its feature set before migrate_params() existed. That's
    why this migration handles two cases.)
    """
    if params["rows"]:
        return {"rows": params["rows"]}
    else:
        return {"rows": "%d-%d" % (params["first_row"], params["last_row"])}


def migrate_params(params):
    if "first_row" in params:
        params = _migrate_params_v0_to_v1(params)
    return params
